File: LegacySource/settlement/room/law/guard/Patrol.cs
using System;
using System.IO;

namespace Settlement.Room.Law.Guard
{
    public sealed class Patrol : SAVABLE
    {
        private const int WIDTH = 1;
        private const int DEPTH = 8;
        public const int MAX = (1 + WIDTH * 2) * DEPTH;
        private const int MAX_TILES = 120;

        public static double Speed = 0.5;

        private readonly PathSimple path = new PathSimple(MAX_TILES + 16);
        private double waitAtDest = RND.rFloat() * TIME.SecondsPerHour() * 3;
        private double progressTile = 0;
        private short[] txs = new short[DEPTH];
        private short[] tys = new short[DEPTH];
        private byte[] dirs = new byte[DEPTH];

        private Coo tmp = new Coo();

        public Patrol()
        {
        }

        public override void Save(FilePutter file)
        {
            file.D(waitAtDest);
            file.D(progressTile);
            file.SsE(txs);
            file.SsE(tys);
            file.BsE(dirs);
        }

        public override void Load(FileGetter file)
        {
            waitAtDest = file.D();
            progressTile = file.D();
            file.SsE(txs);
            file.SsE(tys);
            file.BsE(dirs);
        }

        public override void Clear()
        {
            waitAtDest = 0;
            progressTile = 0;
        }

        public void Update(double ds)
        {
            if (!path.HasNext())
            {
                waitAtDest -= ds;
                if (waitAtDest < 0)
                {
                    Find();

                    waitAtDest = TIME.SecondsPerHour() * (3 + RND.rFloat() * 3);
                }
            }
            else
            {
                progressTile += ds * Speed;

                if (progressTile > 1.0)
                {
                    progressTile -= 1.0;
                    SetNext();
                }
            }
        }

        private void SetNext()
        {
            if (!path.HasNext())
                return;
            path.SetNext();

            for (int i = DEPTH - 1; i > 0; i--)
            {
                txs[i] = txs[i - 1];
                tys[i] = tys[i - 1];
                dirs[i] = dirs[i - 1];
            }

            txs[0] = (short)path.X();
            tys[0] = (short)path.Y();

            int x = path.X();
            int y = path.Y();
            if (path.HasNext())
            {
                path.SetNext();
                DIR dirNext = DIR.Get(x, y, path.X(), path.Y());
                dirs[0] = (byte)dirNext.Id;
                path.SetPrev();
            }
        }

        private void Find()
        {
            int sx = path.X();
            int sy = path.Y();

            if (path.Length() == 0 || !SETT.PATH().Connectivity.Is(path.X(), path.Y()))
            {
                if (SETT.ROOMS().GUARD.InstancesSize() > 0)
                {
                    RoomInstance ins = SETT.ROOMS().GUARD.GetInstance(RND.rInt(SETT.ROOMS().GUARD.InstancesSize()));
                    sx = ins.MX();
                    sy = ins.MY();
                }
                else
                {
                    sx = THRONE.Coo().X;
                    sy = THRONE.Coo().Y;
                    DIR d = DIR.ORTHO.Rnd();
                    sx += d.X;
                    sy += d.Y;
                }
            }

            DIR dirDir = DIR.Get(THRONE.Coo().X, THRONE.Coo().Y, sx, sy);

            Flooder f = GUTIL.Flooder();
            f.Init(this);

            f.PushSloppy(sx, sy, 0);
            f.SetValue2(sx, sy, 0);

            double destDist = MAX_TILES + RND.rInt(MAX_TILES);

            while (f.HasMore())
            {
                PathTile t = f.PollSmallest();

                if (t.GetValue2() > destDist)
                {
                    GUTIL.Coos().Set(0);
                    GUTIL.Coos().Get().Set(t);

                    while (f.HasMore())
                    {
                        t = f.PollSmallest();
                        if (t.GetValue2() > destDist)
                        {
                            GUTIL.Coos().Inc();
                            GUTIL.Coos().Get().Set(t);
                        }
                    }
                    t = GUTIL.PathTools().GetTile(GUTIL.Coos().Get().X, GUTIL.Coos().Get().Y);

                    path.Set(t);

                    for (int i = 0; i < DEPTH; i++)
                        SetNext();
                    f.Done();
                    return;
                }

                for (int di = 0; di < DIR.ALL.Size; di++)
                {
                    DIR d = DIR.ALL.Get(di);
                    int dx = t.X + d.X;
                    int dy = t.Y + d.Y;

                    if (!SETT.IN_BOUNDS(dx, dy))
                        continue;

                    double a = SETT.PATH().Coster.Player.GetCost(t.X, t.Y, dx, dy);
                    if (a < 0)
                        continue;

                    if (SETT.ENV().Map.URBAN.Get(dx, dy) == 0)
                        a *= 5;

                    a *= 10 - 9 * SETT.ENV().Map.SPACE.Get(dx, dy);
                    if (SETT.FLOOR().Getter.Get(dx, dy) == null)
                        a *= 4;
                    if (SETT.ROOMS().Map.Is(dx, dy))
                    {
                        a *= 4;
                    }

                    double dot = d.XN * dirDir.XN + d.YN * dirDir.YN;
                    a += 2 + dot;

                    a *= d.TileDistance();

                    if (f.PushSmaller(dx, dy, t.GetValue + a, t) != null)
                    {
                        f.SetValue2(dx, dy, t.GetValue2 + d.TileDistance());
                    }
                }
            }

            f.Done();
            return;
        }

        public Coo Pos(int pos)
        {
            int side = pos / DEPTH;
            pos = pos % DEPTH;
            DIR d = Dir(pos);

            double tx = txs[pos] + d.X * progressTile;
            double ty = tys[pos] + d.Y * progressTile;

            if (side != 0)
            {
                int depth = (int)Math.Ceiling(side / 2.0);
                d = d.Next(2 + (side % 2) * 4);
                tx += d.XN * (depth);
                ty += d.YN * (depth);
            }
            tmp.Set(tx * C.TILE_SIZE + C.TILE_SIZEH, ty * C.TILE_SIZE + C.TILE_SIZEH);
            return tmp;
        }

        public DIR Dir(int pos)
        {
            pos = pos % DEPTH;
            return DIR.ALL.Get(dirs[pos]);
        }

        public int Posses()
        {
            return MAX;
        }
    }
}

File: settlement/room/law/guard/Patrol.java
package settlement.room.law.guard;

import java.io.IOException;

import game.time.TIME;
import init.constant.C;
import settlement.main.SETT;
import settlement.room.main.RoomInstance;
import settlement.room.main.throne.THRONE;
import snake2d.PathGame.PathSimple;
import snake2d.PathTile;
import snake2d.PathUtilOnline.Flooder;
import snake2d.util.datatypes.Coo;
import snake2d.util.datatypes.DIR;
import snake2d.util.file.Alloc;
import snake2d.util.file.FileGetter;
import snake2d.util.file.FilePutter;
import snake2d.util.file.SAVABLE;
import snake2d.util.rnd.RND;
import util.GUTIL;

public final class Patrol implements SAVABLE {

	private static int WIDTH = 1;
	private static int DEPTH = 8;
	public static int MAX = (1 + WIDTH*2)*DEPTH;
	private static int MAX_TILES = 120;
	
	public static double speed = 0.5;
	
	private final PathSimple path = new PathSimple(MAX_TILES + 16);
	private double waitAtDest = RND.rFloat()*TIME.secondsPerHour()*3;
	private double progressTile = 0;
	private short[] txs = new short[DEPTH];
	private short[] tys = new short[DEPTH];
	private byte[] dirs = Alloc.bb(DEPTH);
	
	private Coo tmp = new Coo();
	
	Patrol(){
		
	}
	
	@Override
	public void save(FilePutter file) {
		file.d(waitAtDest);
		file.d(progressTile);
		file.ssE(txs);
		file.ssE(tys);
		file.bsE(dirs);
	}

	@Override
	public void load(FileGetter file) throws IOException {
		waitAtDest = file.d();
		progressTile = file.d();
		file.ssE(txs);
		file.ssE(tys);
		file.bsE(dirs);
		
	}

	@Override
	public void clear() {
		waitAtDest = 0;
		progressTile = 0;
	}
	
	void update(double ds) {
		
		if (!path.hasNext()) {
			waitAtDest-=ds;
			if (waitAtDest < 0) {
				find();
				
				waitAtDest = TIME.secondsPerHour()*(3+RND.rFloat()*3);
			}else {
			
			}
		}else {
			progressTile += ds*speed;
			
			if (progressTile > 1.0) {
				progressTile-=1.0;
				setNext();
				
			}
		}
	}
	
	private void setNext() {
		
		if (!path.hasNext())
			return;
		path.setNext();
		
		for (int i = DEPTH-1; i > 0; i--) {
			txs[i] = txs[i-1];
			tys[i] = tys[i-1];
			dirs[i] = dirs[i-1];
		}
		
		txs[0] = (short) path.x();
		tys[0] = (short) path.y();
		
		int x = path.x();
		int y = path.y();
		if (path.hasNext()) {
			path.setNext();
			DIR dirNext = DIR.get(x, y, path.x(), path.y());
			dirs[0] = (byte) dirNext.id();
			path.setPrev();
		}
	}
	
	private void find() {
		
		if (false) {
			//Find destinations for all patrols in a single run.
			//impove it generally
			
		}

		int sx = path.x();
		int sy = path.y();
		
		if (path.length() == 0 || !SETT.PATH().connectivity.is(path.x(), path.y())) {
			
			if (false || SETT.ROOMS().GUARD.instancesSize() > 0) {
				RoomInstance ins = SETT.ROOMS().GUARD.getInstance(RND.rInt(SETT.ROOMS().GUARD.instancesSize()));
				sx = ins.mX();
				sy = ins.mY();

			}else {
				sx = THRONE.coo().x();
				sy = THRONE.coo().y();
				DIR d = DIR.ORTHO.rnd();
				sx += d.x();
				sy += d.y();

			}
			
			
		}

		DIR dirDir = DIR.get(THRONE.coo().x(), THRONE.coo().y(), sx, sy);
		
		Flooder f = GUTIL.flooder();
		f.init(this);
		
		f.pushSloppy(sx, sy, 0);
		f.setValue2(sx, sy, 0);
		
		double destDist = MAX_TILES + RND.rInt(MAX_TILES);
		
		while(f.hasMore()) {
			
			PathTile t = f.pollSmallest();
			
			if (t.getValue2() > destDist) {
				
				GUTIL.coos().set(0);
				GUTIL.coos().get().set(t);
				
				while(f.hasMore()) {
					t = f.pollSmallest();
					if (t.getValue2() > destDist) {
						GUTIL.coos().inc();
						GUTIL.coos().get().set(t);
					}
				}
				t = GUTIL.pathTools().getTile(GUTIL.coos().get().x(), GUTIL.coos().get().y());

				path.set(t);

				for (int i = 0; i < DEPTH; i++)
					setNext();
				f.done();
				return;
			}
			
			for (int di = 0; di < DIR.ALL.size(); di++) {
				DIR d = DIR.ALL.get(di);
				int dx = t.x() + d.x();
				int dy = t.y() + d.y();
				
				if (!SETT.IN_BOUNDS(dx, dy)) {
					continue;
				}
				
				double a = SETT.PATH().coster.player.getCost(t.x(), t.y(), dx, dy);
				if (a < 0)
					continue;
				
				if (SETT.ENV().map.URBAN.get(dx, dy) == 0)
					a *= 5;
				
				a *= 10 - 9*SETT.ENV().map.SPACE.get(dx, dy);
				if (SETT.FLOOR().getter.get(dx, dy) == null)
					a *= 4;
				if (SETT.ROOMS().map.is(dx, dy)) {
					a *= 4;
				}
				
				double dot = d.xN()*dirDir.xN() + d.yN()*dirDir.yN();
				a += 2 + dot;
				
				a*=d.tileDistance();
				
				if (f.pushSmaller(dx, dy, t.getValue()+a, t) != null) {
					f.setValue2(dx, dy, t.getValue2()+d.tileDistance());
				}
				
			}
			
			
		}

		f.done();
		return;
		
		
	}
	
	public Coo pos(int pos) {
		int side = pos/DEPTH;
		pos = pos%DEPTH;
		DIR d = dir(pos);
		
		double tx = txs[pos] + d.x()*progressTile;
		double ty = tys[pos] + d.y()*progressTile;
		
		if (side != 0) {
			int depth = (int) Math.ceil(side/2.0);
			d = d.next(2 + (side%2)*4);
			tx += d.xN()*(depth);
			ty += d.yN()*(depth);
		}
		tmp.set(tx*C.TILE_SIZE + C.TILE_SIZEH, ty*C.TILE_SIZE + C.TILE_SIZEH);
		return tmp;
		
	}
	
	public DIR dir(int pos) {
		pos = pos%DEPTH;
		return DIR.ALL.get(dirs[pos]);
	}
	
	public int posses() {
		return MAX;
	}


	
}
